packages/shared/ReactErrorUtils.js

```diff
@@ -8,110 +8,117 @@
  */
 
 import invariant from 'shared/invariant';
-import invokeGuardedCallback from './invokeGuardedCallback';
+import invokeGuardedCallbackImpl from './invokeGuardedCallbackImpl';
 
-const ReactErrorUtils = {
-  // Used by Fiber to simulate a try-catch.
-  _caughtError: (null: mixed),
-  _hasCaughtError: (false: boolean),
+// Used by Fiber to simulate a try-catch.
+let hasError: boolean = false;
+let caughtError: mixed = null;
 
-  // Used by event system to capture/rethrow the first error.
-  _rethrowError: (null: mixed),
-  _hasRethrowError: (false: boolean),
+// Used by event system to capture/rethrow the first error.
+let hasRethrowError: boolean = false;
+let rethrowError: mixed = null;
 
-  /**
-   * Call a function while guarding against errors that happens within it.
-   * Returns an error if it throws, otherwise null.
-   *
-   * In production, this is implemented using a try-catch. The reason we don't
-   * use a try-catch directly is so that we can swap out a different
-   * implementation in DEV mode.
-   *
-   * @param {String} name of the guard to use for logging or debugging
-   * @param {Function} func The function to invoke
-   * @param {*} context The context to use when calling the function
-   * @param {...*} args Arguments for function
-   */
-  invokeGuardedCallback: function<A, B, C, D, E, F, Context>(
-    name: string | null,
-    func: (a: A, b: B, c: C, d: D, e: E, f: F) => mixed,
-    context: Context,
-    a: A,
-    b: B,
-    c: C,
-    d: D,
-    e: E,
-    f: F,
-  ): void {
-    invokeGuardedCallback.apply(ReactErrorUtils, arguments);
-  },
-
-  /**
-   * Same as invokeGuardedCallback, but instead of returning an error, it stores
-   * it in a global so it can be rethrown by `rethrowCaughtError` later.
-   * TODO: See if _caughtError and _rethrowError can be unified.
-   *
-   * @param {String} name of the guard to use for logging or debugging
-   * @param {Function} func The function to invoke
-   * @param {*} context The context to use when calling the function
-   * @param {...*} args Arguments for function
-   */
-  invokeGuardedCallbackAndCatchFirstError: function<A, B, C, D, E, F, Context>(
-    name: string | null,
-    func: (a: A, b: B, c: C, d: D, e: E, f: F) => void,
-    context: Context,
-    a: A,
-    b: B,
-    c: C,
-    d: D,
-    e: E,
-    f: F,
-  ): void {
-    ReactErrorUtils.invokeGuardedCallback.apply(this, arguments);
-    if (ReactErrorUtils.hasCaughtError()) {
-      const error = ReactErrorUtils.clearCaughtError();
-      if (!ReactErrorUtils._hasRethrowError) {
-        ReactErrorUtils._hasRethrowError = true;
-        ReactErrorUtils._rethrowError = error;
-      }
-    }
-  },
-
-  /**
-   * During execution of guarded functions we will capture the first error which
-   * we will rethrow to be handled by the top level error handler.
-   */
-  rethrowCaughtError: function() {
-    return rethrowCaughtError.apply(ReactErrorUtils, arguments);
+const reporter = {
```

Comment:
We can technically get rid of this object and just pass `onError` as context directly. Not sure if it looks too confusing.

Reply:
I also don’t want to make it an extra argument to avoid having to mess with `arguments` when we forward the calls.